File: bindings/php/src/Core/Nodes/Flags.php
<?php

namespace STRling\Core\Nodes;

use JsonSerializable;


// ---- Flags ----

class Flags implements JsonSerializable
{
    public function __construct(
        public bool $ignoreCase = false,
        public bool $multiline = false,
        public bool $dotAll = false,
        public bool $unicode = false,
        public bool $extended = false,
    ) {}

    public static function fromLetters(string $letters): self
    {
        $f = new self();
        $cleaned = str_replace([',', ' '], '', $letters);
        
        foreach (str_split($cleaned) as $ch) {
            match ($ch) {
                'i' => $f->ignoreCase = true,
                'm' => $f->multiline = true,
                's' => $f->dotAll = true,
                'u' => $f->unicode = true,
                'x' => $f->extended = true,
                default => null,
            };
        }
        
        return $f;
    }

    public function jsonSerialize(): mixed
    {
        return [
            'ignoreCase' => $this->ignoreCase,
            'multiline' => $this->multiline,
            'dotAll' => $this->dotAll,
            'unicode' => $this->unicode,
            'extended' => $this->extended,
        ];
    }
}

File: bindings/fsharp/src/STRling/Core/Flags.fs
namespace STRling.Core

/// Container for regex flags/modifiers.
type Flags = {
    IgnoreCase: bool
    Multiline: bool
    DotAll: bool
    Unicode: bool
    Extended: bool
}

module Flags =
    /// Default flags with all options disabled.
    let defaultFlags = {
        IgnoreCase = false
        Multiline = false
        DotAll = false
        Unicode = false
        Extended = false
    }

    /// Creates Flags from a string of flag letters (e.g., "imsu").
    let fromLetters (letters: string) =
        let cleaned = letters.Replace(",", "").Replace(" ", "").ToLower()
        let mutable ignoreCase = false
        let mutable multiline = false
        let mutable dotAll = false
        let mutable unicode = false
        let mutable extended = false

        for ch in cleaned do
            match ch with
            | 'i' -> ignoreCase <- true
            | 'm' -> multiline <- true
            | 's' -> dotAll <- true
            | 'u' -> unicode <- true
            | 'x' -> extended <- true
            | _ -> ()

        { IgnoreCase = ignoreCase
          Multiline = multiline
          DotAll = dotAll
          Unicode = unicode
          Extended = extended }

    /// Convert flags to a dictionary for serialization.
    let toDict (flags: Flags) =
        dict [
            "ignoreCase", flags.IgnoreCase
            "multiline", flags.Multiline
            "dotAll", flags.DotAll
            "unicode", flags.Unicode
            "extended", flags.Extended
        ]
